### ISSUE TITLE
Discrepancy between Conv1D implementation of FbGemm vs LibTorch/PyTorch

### ISSUE THREAD

**Issue by abhinavkulkarni** [facebookresearch/wav2letter#807]
### `Conv1D` logic is implemented differently in fbgemm vs LibTorch/PyTorch/TensorFlow/Keras

TL;DR: FbGemm implementation of `Conv1D` keeps only one weight matrix overall vs one weight matrix per group

&nbsp;
&nbsp;

`Conv1D` layer takes a `N` `inChannels`-dimensional vectors as input and produces `M` `outChannels`-dimensional vectors as output. The value of `M` is determined by `N`, `kernelSize`, `stride` and  `padding`.

For e.g., in [this](https://missinglink.ai/wp-content/uploads/2019/03/1D-convolutional-example_2x.png) picture, 

```cpp
N=9, inChannels=6, kernelSize=2, stride=1, padding=0, M=8
```

The weight matrix is of the size `outChannels x inChannels x kernelSize`.

There is an additional `groups` parameter. When `groups > 1`, there are `groups` number of weight matrices. Each matrix is of the size `outChannels/groups x inChannels/groups x kernelSize`. Essentially each matrix is responsible for convolving `groups` share of input channels to produce `groups` share of output channels.

So, after concatenating these `groups` number of weight matrices, the total size of the final weight matrix is `outChannels x inChannels/groups x kernelSize`. This helps in reducing the number of parameters (while sacrificing little to nothing on accuracy).

Here is an PyTorch example:

```python
>>> import torch
>>> conv1d = torch.nn.Conv1d(in_channels=4, out_channels=16, kernel_size=3, stride=2, padding=0, groups=2)
>>> conv1d.weight.shape
torch.Size([16, 2, 3])
```

However, in FbGemm implementation, only 1 copy of the weight matrix is kept overall (instead of one copy per group) and thus the size of the overall weight matrix is `outChannels/groups x inChannels/groups x kernelSize` as can be seen [here](https://github.com/facebookresearch/wav2letter/blob/v0.2/inference/inference/module/nn/backend/fbgemm/Conv1dFbGemm.cpp#L59).

Is it intentional to keep only one copy of the weight matrix and not different copies per group? Does this discrepancy exist because ArrayFire implementation of `Conv1D` keeps only one copy of the weight matrix overall (vs one per group)?

Thanks!

**Comment by abhinavkulkarni** [comment]
Tagging @tlikhomanenko 

Thanks!

**Comment by vineelpratap** [comment]
Hi, 
  Thanks for mentioning this. Yes, there seems to be a bug in the notation in inference pipeline and needs to be fixed. 
  However, note that it won't affect the correctness of the current streaming ASR pipeline and we can create `Conv1dFbGemm ` layer with `outChannels(new) = outChannels(old) / groups` .

**Comment by abhinavkulkarni** [comment]
@vineelpratap: Thanks for responding.

Can you please elaborate more? Specifically,

1. What exactly is happening currently during inference? Whenever `Conv1D` module is being used (by itself or as part of larger building block like a `TDSBlock`, it keeps only one copy of weight matrix right (as opposed to one weight matrix per group)? I am currently in the process of replacing FbGemm backend with LibTorch and for now I am going to replicate this behavior.

2. As far as I understood, only `Linear` and `Conv1d` layers need to be reimplemented in a new backend for the inference pipeline. As long as I correctly mimic the existing FbGemm behavior, it should be possible for me to use my own custom backend.

Let me know.

Thanks

**Comment by vineelpratap** [comment]
For example, if you are trying to replace the layer `Conv1dFbGemm(inChannels, outChannels, kernel, groups)`, you can create `torch.conv1d(inChannels, outChannels/groups, kernel, groups)` - they should be equivalent. Note that weight you might have to do `.permute` to make sure the data ordering is correct (I'm not sure). 

> only Linear and Conv1d layers need to be reimplemented in a new backend for the inference

Yes, that's true. And you make it work on on ANY device. Note that if you trying to replace `FbGemm` code, you are in fact to trying to replace a matrix multiplication. So, there is no need to bother about channel/group things. 
For example, the code [here](https://github.com/facebookresearch/wav2letter/blob/v0.2/inference/inference/module/nn/backend/fbgemm/Conv1dFbGemm.cpp#L59) is doing matrix multiplication with `inputChannels = (inChannels_ / groups_) * kernelSize_` and `outputChannels=(outChannels_ / groups_)` (may be with some `transpose` involved). So, as long as you can replicate this behavior it should be fine.

**Comment by abhinavkulkarni** [comment]
Thanks, @vineelpratap, I was able to replicate the behavior in `LibTorch`. Closing this issue for now.
